src/librustc/dep_graph/shadow.rs

```diff
@@ -0,0 +1,123 @@
+//! The "Shadow Graph" is maintained on the main thread and which
+//! tracks each message relating to the dep-graph and applies some
+//! sanity checks as they go by. If an error results, it means you get
+//! a nice stack-trace telling you precisely what caused the error.
+//!
+//! NOTE: This is a debugging facility which can potentially have non-trivial
+//! runtime impact. Therefore, it is largely compiled out if
+//! debug-assertions are not enabled.
+//!
+//! The basic sanity check, always enabled, is that there is always a
+//! task (or ignore) on the stack when you do read/write.
```

Comment:
> But it looks like control flow can never reach check_edge if debug assertions are turned off?

By "always enabled", I meant "always enabled if you build w/ debug assertions" (you don't need to e.g. use an env variable)